### ISSUE TITLE
3.5 Why is thread-safety of Subscription.cancel() required?

### ISSUE THREAD

**Issue by anthonyvdotbe** [reactive-streams/reactive-streams-jvm#319]
Rule 3.5 says that Subscription.cancel() MUST be thread-safe.
Why is this so, given that rule 2.7 excludes the possibility of Subscription.cancel() being invoked concurrently? Why isn't it sufficient if the fact of cancellation is guaranteed to be visible to any subsequent calls on the Subscription?

For example, say I have something like the following:

```
volatile boolean cancelled = false;

public void cancel() {
    myHashMap.put("hello", "world");
    cancelled = true;
}
```

Why isn't this allowed?

**Comment by akarnokd** [comment]
Experience with RxJava 2, Rsc and Reactor indicates that rule 2.7 adds unnecessary burden because cancels can happen from any thread and in any numbers. For example, a timeout cancels a stream that is cancelled by a limiting operator at the same time.

In contrast, making `cancel` thread safe is usually very simple, just like in your example. It usually sets a [cancelled flag](https://github.com/reactor/reactive-streams-commons/blob/master/src/main/java/reactivestreams/commons/publisher/PublisherRange.java#L95), or [performs some CAS that succeeds only once](https://github.com/reactor/reactive-streams-commons/blob/master/src/main/java/reactivestreams/commons/publisher/PublisherWindow.java#L240) and does cleanup, [etc](https://github.com/reactor/reactive-streams-commons/blob/master/src/main/java/reactivestreams/commons/publisher/PublisherObserveOn.java#L262).

**Comment by viktorklang** [comment]
@akarnokd Why would it be a burden, all which is required is to wrap the original Subscription in a safe wrapper? Or did I misunderstand?

**Comment by akarnokd** [comment]
It's not only the wrapping but the serialization primitive overhead. Plus the hundreds of operators I've written, the thread-safe cancellation was an easy / trivial thing (see linked by the examples).

**Comment by viktorklang** [comment]
@akarnokd Yes, threadsafe cancellation is "easy", however, should the spec mandate how it is implemented? Or what do you propose given your experience?

**Comment by akarnokd** [comment]
Drop §2.7. §3.5 is fine by itself.

**Comment by viktorklang** [comment]
@akarnokd §2.7 covers more than `cancel` though.

**Comment by viktorklang** [comment]
AFAIR addressed by #339
